### PR TITLE
executor: do not return first row until the frame is completed.

### LINKED ISSUE #12415: window function panic: runtime error: index out of range
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
SELECT
	income.customer_id,
	income.contract_id,
	income.pay_time,
	first_value ( contract_id ) over ( PARTITION BY customer_id ORDER BY pay_time ASC ) AS first_contract_id 
FROM
	contract_tidb.finance_income income

2. What did you expect to see?
     return columns

3. What did you see instead?
    **tidb.log  error:**  
   [ERROR] [conn.go:600] ["connection running loop panic"] [conn=5664] [lastCmd="SELECT
	income.customer_id,
	income.contract_id,
	income.pay_time,
	first_value ( contract_id ) over ( PARTITION BY customer_id ORDER BY pay_time ASC ) AS first_contract_id 
FROM
	contract_tidb.finance_income income"] [err="\"index out of range\""] [stack="goroutine 15400945 [running]:
github.com/pingcap/tidb/server.(*clientConn).Run.func1(0x227cee0, 0xc013c58840, 0xc01d4dad00)
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:598 +0xee
panic(0x1daad60, 0x31d59f0)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/pingcap/tidb/server.(*clientConn).writeResultset.func1(0x0, 0x228ece0, 0xc00adfb590, 0xc02e2cd738, 0x227cee0, 0xc013c58840, 0xc01d4dad00)
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:1261 +0x59c
panic(0x1daad60, 0x31d59f0)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/pingcap/tidb/util/chunk.(*column).isNull(...)
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/util/chunk/column.go:71
github.com/pingcap/tidb/util/chunk.Row.IsNull(...)
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/util/chunk/row.go:223
github.com/pingcap/tidb/expression.(*Column).EvalInt(0xc01064bd10, 0x22b63e0, 0xc02eddd0e0, 0xc001bed080, 0x3ff, 0xc02eddd0e0, 0x22b76a0, 0xc01064be00, 0x22b76a0)
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/expression/column.go:221 +0x267
github.com/pingcap/tidb/executor/aggfuncs.(*value4Int).extractRow(0xc00b7ab880, 0x22b63e0, 0xc02eddd0e0, 0x22b76a0, 0xc01064bd10, 0xc001bed080, 0x3ff, 0x0, 0x1)
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/aggfuncs/func_value.go:42 +0x63
github.com/pingcap/tidb/executor/aggfuncs.(*firstValue).UpdatePartialResult(0xc001becb10, 0x22b63e0, 0xc02eddd0e0, 0xc013952000, 0x2, 0x80, 0xc000f327c0, 0x0, 0x0)
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/aggfuncs/func_value.go:247 +0x9f
github.com/pingcap/tidb/executor.(*rangeFrameWindowProcessor).appendResult2Chunk(0xc022dd1580, 0x22b63e0, 0xc02eddd0e0, 0xc013952000, 0x2, 0x80, 0xc001becc60, 0x1, 0x1ea8680, 0x0, ...)
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/window.go:404 +0x377
github.com/pingcap/tidb/executor.(*WindowExec).appendResult2Chunk(0xc0015542d0, 0xc001becc60, 0x0, 0x0)
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/window.go:149 +0xc1
github.com/pingcap/tidb/executor.(*WindowExec).Next(0xc0015542d0, 0x227cee0, 0xc013c58840, 0xc001becc60, 0x0, 0x0)
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/window.go:66 +0x1df
github.com/pingcap/tidb/executor.Next(0x227cee0, 0xc013c58840, 0x2283f60, 0xc0015542d0, 0xc001becc60, 0x23, 0x13)
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/executor.go:191 +0xbd
github.com/pingcap/tidb/executor.(*recordSet).Next(0xc00adfb540, 0x227cee0, 0xc013c58840, 0xc001becc60, 0x0, 0x0)
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/adapter.go:109 +0xb0
github.com/pingcap/tidb/server.(*tidbResultSet).Next(0xc00adfb590, 0x227cee0, 0xc013c58840, 0xc001becc60, 0x0, 0x0)
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/driver_tidb.go:365 +0x51
github.com/pingcap/tidb/server.(*clientConn).writeChunks(0xc01d4dad00, 0x227cee0, 0xc013c58840, 0x228ece0, 0xc00adfb590, 0xc00000d700, 0x227cee0, 0xc013c58840)
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:1308 +0x319
github.com/pingcap/tidb/server.(*clientConn).writeResultset(0xc01d4dad00, 0x227cee0, 0xc013c58840, 0x228ece0, 0xc00adfb590, 0xc000005800, 0x0, 0x0, 0x0)
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:1274 +0x1a1
github.com/pingcap/tidb/server.(*clientConn).handleQuery(0xc01d4dad00, 0x227cee0, 0xc013c58840, 0xc0014e0fc1, 0xd7, 0x0, 0x0)
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:1191 +0x212
github.com/pingcap/tidb/server.(*clientConn).dispatch(0xc01d4dad00, 0x227cee0, 0xc013c58840, 0xc0014e0fc1, 0xd8, 0xd8, 0x0, 0x0)
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:897 +0x5c6
github.com/pingcap/"] [stack="github.com/pingcap/tidb/server.(*clientConn).Run.func1
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:600
runtime.gopanic
	/usr/local/go/src/runtime/panic.go:522
github.com/pingcap/tidb/server.(*clientConn).writeResultset.func1
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:1261
runtime.gopanic
	/usr/local/go/src/runtime/panic.go:522
runtime.panicindex
	/usr/local/go/src/runtime/panic.go:44
github.com/pingcap/tidb/util/chunk.(*column).isNull
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/util/chunk/column.go:71
github.com/pingcap/tidb/util/chunk.Row.IsNull
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/util/chunk/row.go:223
github.com/pingcap/tidb/expression.(*Column).EvalInt
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/expression/column.go:221
github.com/pingcap/tidb/executor/aggfuncs.(*value4Int).extractRow
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/aggfuncs/func_value.go:42
github.com/pingcap/tidb/executor/aggfuncs.(*firstValue).UpdatePartialResult
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/aggfuncs/func_value.go:247
github.com/pingcap/tidb/executor.(*rangeFrameWindowProcessor).appendResult2Chunk
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/window.go:404
github.com/pingcap/tidb/executor.(*WindowExec).appendResult2Chunk
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/window.go:149
github.com/pingcap/tidb/executor.(*WindowExec).Next
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/window.go:66
github.com/pingcap/tidb/executor.Next
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/executor.go:191
github.com/pingcap/tidb/executor.(*recordSet).Next
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/adapter.go:109
github.com/pingcap/tidb/server.(*tidbResultSet).Next
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/driver_tidb.go:365
github.com/pingcap/tidb/server.(*clientConn).writeChunks
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:1308
github.com/pingcap/tidb/server.(*clientConn).writeResultset
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:1274
github.com/pingcap/tidb/server.(*clientConn).handleQuery
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:1191
github.com/pingcap/tidb/server.(*clientConn).dispatch
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:897
github.com/pingcap/tidb/server.(*clientConn).Run
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:652
github.com/pingcap/tidb/server.(*Server).onConn
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/server.go:440"]
[2019/09/26 14:11:50.344 +08:00] [INFO] [server.go:416] ["connection closed"] [conn=5664]

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
Release Version: v3.0.3
Git Commit Hash: 836982c617fbaa42d74616eb9ba9ed4cae46c46b
Git Branch: HEAD
UTC Build Time: 2019-08-30 02:42:42
GoVersion: go version go1.12 linux/amd64
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false

### PR DESCRIPTION
**Pull request by SunRunAway**
### What problem does this PR solve? 

Fixes #12415.

### What is changed and how it works?

In previous implementation, `windowExec` returns a chunk when a frame is fetched but not evaluated. After rewriting, `windowExec` returns a chunk after a frame is finished evaluating.

It makes the code simpler, and all the implementation of aggregation functions can reference the data in the frame. The bad side is it will increase the time of meeting first row.

I'll send a following PR(#12481) to revert the changes of copying data in #11678, if this is accepted by reviewers.

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below) 
    - the case in #12415 passes.

Code changes

 - None

Side effects

 - It will increase the time of meeting first row

Related changes

 - None

Release note

 - Fix a memory reference bug in window function.

### CODE REVIEW

File: executor/window_test.go [hunk 1]
```diff
@@ -20,6 +20,7 @@ import (
 
 func (s *testSuite4) TestWindowFunctions(c *C) {
 	tk := testkit.NewTestKit(c, s.store)
+	var result *testkit.Result
```
Comment (lamxTyler) [inline executor/window_test.go]: Why change it?

File: executor/window.go [hunk 2]
```diff
@@ -30,137 +30,138 @@ import (
...
-	remainingRowsInGroup int
-	remainingRowsInChunk int
-	numWindowFuncs       int
-	processor            windowProcessor
+	groupChecker *groupChecker
+	// inputIter is the iterator of children chunks
+	inputIter *chunk.Iterator4Chunk
+	// executed indicates the child executor is drained or something unexpected happened.
+	executed     bool
+	requiredRows int
```
Comment (lamxTyler) [inline executor/window.go]: Why we need it?

File: executor/window.go [hunk 3]
```diff
@@ -30,137 +30,138 @@ import (
...
-	e.groupRows, err = e.processor.appendResult2Chunk(e.ctx, e.groupRows, chk, remained)
-	if err != nil {
-		return err
-	}
-	e.remainingRowsInGroup -= remained
-	e.remainingRowsInChunk -= remained
-	if e.remainingRowsInGroup == 0 {
-		e.processor.resetPartialResult()
-		e.groupRows = e.groupRows[:0]
+	resultChk := chunk.New(e.retFieldTypes, 0, e.requiredRows)
```
Comment (lamxTyler) [inline executor/window.go]: Why not `numRows`?

File: executor/window.go [hunk 4]
```diff
@@ -30,137 +30,138 @@ import (
...
 
 // Close implements the Executor Close interface.
 func (e *WindowExec) Close() error {
-	e.childResults = nil
 	return errors.Trace(e.baseExecutor.Close())
 }
 
 // Next implements the Executor Next interface.
 func (e *WindowExec) Next(ctx context.Context, chk *chunk.Chunk) error {
+	if e.requiredRows == 0 { // we assume that each future chk has the same RequiredRows.
```
Comment (lamxTyler) [inline executor/window.go]: Comment should start with uppercase.

File: executor/window.go [hunk 5]
```diff
@@ -30,137 +30,138 @@ import (
...
 	}
-	e.groupRows, err = e.processor.consumeGroupRows(e.ctx, e.groupRows)
-	if err != nil {
-		return errors.Trace(err)
+	for i := 0; i < len(e.resultChunk); i++ {
+		remained := mathutil.Min(e.remainingRowsInChunk[i], remainingRowsInGroup)
+		e.remainingRowsInChunk[i] -= remained
+		remainingRowsInGroup -= remained
+
+		// TODO: combine these three methods
```
Comment (lamxTyler) [inline executor/window.go]: Ditto.

File: executor/window_test.go (hunk 1 above)
Comment (SunRunAway) [reply executor/window_test.go]: Make debug using comment more easily.

File: executor/window.go [hunk 6]
```diff
@@ -30,137 +30,133 @@ import (
...
-	if e.inputIter != nil && e.inputRow != e.inputIter.End() {
-		return nil
-	}
-
-	// Before fetching a new batch of input, we should consume the last group rows.
-	err = e.consumeGroupRows()
-	if err != nil {
-		return errors.Trace(err)
+func (e *WindowExec) fetchChildIfNecessary(ctx context.Context) (EOF bool, err error) {
+	if e.inputIter != nil && e.inputIter.Current() != e.inputIter.End() {
```
Comment (lzmhhh123) [inline executor/window.go]: Is any possibility to get into the branch?

File: executor/window.go [hunk 7]
```diff
@@ -30,137 +30,133 @@ import (
...
-func (e *WindowExec) appendResult2Chunk(chk *chunk.Chunk) (err error) {
-	if err := e.copyChk(chk); err != nil {
-		return err
-	}
-	remained := mathutil.Min(e.remainingRowsInChunk, e.remainingRowsInGroup)
-	e.groupRows, err = e.processor.appendResult2Chunk(e.ctx, e.groupRows, chk, remained)
-	if err != nil {
-		return err
+	resultChk := chunk.New(e.retFieldTypes, 0, numRows)
+	if err := e.copyChk(childResult, resultChk); err != nil {
```
Comment (lzmhhh123) [inline executor/window.go]:
```suggestion
	err := e.copyChk(childResult, resultChk)
	if err != nil {
```

File: executor/window.go (hunk 6 above)
Comment (SunRunAway) [reply executor/window.go]: When a `consumeOneGroup` exits because `meetNewGroup`, the next `consumeOneGroup` must be starts with this branch.

**Pull request merged**